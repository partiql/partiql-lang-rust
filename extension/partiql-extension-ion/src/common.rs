/// The encoding to use when decoding/encoding Ion to/from PartiQL [`partiql_value::Value`]
#[derive(Copy, Clone, Eq, PartialEq)]
pub enum Encoding {
    /// 'Unlifted'/'Unlowered' Ion to/from PartiQL [`partiql_value::Value`]. [`partiql_value::Value`]s that do not have a direct
    /// Ion analog will result in an error (e.g. PartiQL [`partiql_value::Value`] has a 'bag' type, but Ion does not,
    /// so attempting to encode a 'bag' results in an error).
    Ion,
    /// PartiQL encoded into Ion as supported by `partiql-tests` and by other implementations.
    /// (e.g., a PartiQL bag is encoded as an Ion list annotated with "$bag").
    PartiqlEncodedAsIon,
}

pub(crate) const BAG_ANNOT: &str = "$bag";
pub(crate) const TIME_ANNOT: &str = "$time";
pub(crate) const DATE_ANNOT: &str = "$date";
pub(crate) const MISSING_ANNOT: &str = "$missing";

pub(crate) const TIME_PART_HOUR_KEY: &str = "hour";
pub(crate) const TIME_PART_MINUTE_KEY: &str = "minute";
pub(crate) const TIME_PART_SECOND_KEY: &str = "second";
pub(crate) const TIME_PART_TZ_HOUR_KEY: &str = "timezone_hour";
pub(crate) const TIME_PART_TZ_MINUTE_KEY: &str = "timezone_minute";

pub(crate) const RE_SET_TIME_PARTS: [&str; 5] = [
    "^hour$",
    "^minute$",
    "^second$",
    "^timezone_hour$",
    "^timezone_minute$",
];
pub(crate) const TIME_PARTS_HOUR: usize = 0;
pub(crate) const TIME_PARTS_MINUTE: usize = 1;
pub(crate) const TIME_PARTS_SECOND: usize = 2;
pub(crate) const TIME_PARTS_TZ_HOUR: usize = 3;
pub(crate) const TIME_PARTS_TZ_MINUTE: usize = 4;
