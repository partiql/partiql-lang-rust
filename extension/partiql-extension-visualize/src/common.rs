pub trait ToDotGraph<T> {
    fn to_graph(self, data: &T) -> String;
}

#[cfg(feature = "visualize-dot")]
pub(crate) const FG_COLOR: &str = "\"#839496\"";
