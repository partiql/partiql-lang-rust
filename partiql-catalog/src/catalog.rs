use crate::call_defs::ScalarFnCallSpecs;
use crate::scalar_fn::ScalarFunction;
use crate::table_fn::TableFunction;
use partiql_common::catalog::{CatalogId, EntryId, ObjectId};
use partiql_types::PartiqlShape;
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::atomic::{AtomicU64, Ordering};
use thiserror::Error;
use unicase::UniCase;

/// Contains the errors that occur during Catalog related operations
#[derive(Error, Debug, Clone, PartialEq)]
#[error("Catalog error: encountered errors")]
pub struct CatalogError {
    pub errors: Vec<CatalogErrorKind>,
}

impl CatalogError {
    #[must_use]
    pub fn new(errors: Vec<CatalogErrorKind>) -> Self {
        CatalogError { errors }
    }
}

/// Catalog Error kind
///
/// ### Notes
/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
#[derive(Error, Debug, Clone, PartialEq)]
#[non_exhaustive]
pub enum CatalogErrorKind {
    /// Entry exists error.
    #[error("Catalog error: entry already exists for `{0}`")]
    EntryExists(String),

    /// Entry error.
    #[error("Catalog error: `{0}`")]
    EntryError(String),

    /// Any other catalog error.
    #[error("Catalog error: unknown error")]
    Unknown,
}

pub trait Catalog: Debug {
    fn add_table_function(&mut self, info: TableFunction) -> Result<ObjectId, CatalogError>;
    fn add_scalar_function(&mut self, info: ScalarFunction) -> Result<ObjectId, CatalogError>;

    fn add_type_entry(&mut self, entry: TypeEnvEntry<'_>) -> Result<ObjectId, CatalogError>;

    fn get_function(&self, name: &str) -> Option<FunctionEntry<'_>>;

    fn get_function_by_id(&self, id: ObjectId) -> Option<FunctionEntry<'_>>;

    fn resolve_type(&self, name: &str) -> Option<TypeEntry>;
}

#[derive(Debug)]
pub struct TypeEnvEntry<'a> {
    name: UniCase<String>,
    aliases: Vec<&'a str>,
    ty: PartiqlShape,
}

impl<'a> TypeEnvEntry<'a> {
    #[must_use]
    pub fn new(name: &str, aliases: &[&'a str], ty: PartiqlShape) -> Self {
        TypeEnvEntry {
            name: UniCase::from(name.to_string()),
            aliases: aliases.to_vec(),
            ty,
        }
    }
}

#[derive(Debug, Clone)]
pub struct TypeEntry {
    id: ObjectId,
    ty: PartiqlShape,
}

impl TypeEntry {
    #[must_use]
    pub fn id(&self) -> &ObjectId {
        &self.id
    }

    #[must_use]
    pub fn ty(&self) -> &PartiqlShape {
        &self.ty
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct FunctionEntry<'a> {
    id: ObjectId,
    function: &'a FunctionEntryFunction,
}

#[derive(Debug)]
pub enum FunctionEntryFunction {
    Table(TableFunction),
    Scalar(ScalarFnCallSpecs),
    Aggregate(),
}

impl<'a> FunctionEntry<'a> {
    pub fn id(&self) -> &ObjectId {
        &self.id
    }

    #[must_use]
    pub fn entry(&'a self) -> &'a FunctionEntryFunction {
        self.function
    }
}

#[derive(Debug)]
pub struct PartiqlCatalog {
    functions: CatalogEntrySet<FunctionEntryFunction>,
    types: CatalogEntrySet<PartiqlShape>,
    id: CatalogId,
}

impl Default for PartiqlCatalog {
    fn default() -> Self {
        PartiqlCatalog {
            functions: Default::default(),
            types: Default::default(),
            id: 1.into(),
        }
    }
}

impl PartiqlCatalog {}

impl Catalog for PartiqlCatalog {
    fn add_table_function(&mut self, info: TableFunction) -> Result<ObjectId, CatalogError> {
        let call_def = info.call_def();
        let names = call_def.names.clone();
        if let Some((name, aliases)) = names.split_first() {
            let eid = self
                .functions
                .add(name, aliases, FunctionEntryFunction::Table(info))?;
            Ok((self.id, eid).into())
        } else {
            Err(CatalogError::new(vec![CatalogErrorKind::EntryError(
                "Function definition has no name".into(),
            )]))
        }
    }

    fn add_scalar_function(&mut self, info: ScalarFunction) -> Result<ObjectId, CatalogError> {
        let id = self.id;
        let call_def = info.into_call_def();
        let names = call_def.names;
        if let Some((name, aliases)) = names.split_first() {
            self.functions
                .add(
                    name,
                    aliases,
                    FunctionEntryFunction::Scalar(call_def.overloads),
                )
                .map(|eid| ObjectId::new(id, eid))
        } else {
            Err(CatalogError::new(vec![CatalogErrorKind::EntryError(
                "Function definition has no name".into(),
            )]))
        }
    }

    fn add_type_entry(&mut self, entry: TypeEnvEntry<'_>) -> Result<ObjectId, CatalogError> {
        let eid = self
            .types
            .add(entry.name.as_ref(), entry.aliases.as_slice(), entry.ty);

        match eid {
            Ok(eid) => Ok((self.id, eid).into()),
            Err(e) => Err(e),
        }
    }

    fn get_function(&self, name: &str) -> Option<FunctionEntry<'_>> {
        self.functions
            .find_by_name(name)
            .map(|(e, f)| self.to_function_entry(e, f))
    }

    fn get_function_by_id(&self, id: ObjectId) -> Option<FunctionEntry<'_>> {
        assert_eq!(self.id, id.catalog_id());
        self.functions
            .find_by_id(&id.entry_id())
            .map(|(e, f)| self.to_function_entry(e, f))
    }

    fn resolve_type(&self, name: &str) -> Option<TypeEntry> {
        self.types.find_by_name(name).map(|(eid, entry)| TypeEntry {
            id: (self.id, eid).into(),
            ty: entry.clone(),
        })
    }
}

impl PartiqlCatalog {
    fn to_function_entry<'a>(
        &'a self,
        eid: EntryId,
        entry: &'a FunctionEntryFunction,
    ) -> FunctionEntry<'a> {
        FunctionEntry {
            id: (self.id, eid).into(),
            function: entry,
        }
    }
}

#[derive(Debug)]
struct CatalogEntrySet<T> {
    entries: HashMap<EntryId, T>,
    by_name: HashMap<UniCase<String>, EntryId>,
    by_alias: HashMap<UniCase<String>, EntryId>,

    next_id: AtomicU64,
}

impl<T> Default for CatalogEntrySet<T> {
    fn default() -> Self {
        CatalogEntrySet {
            entries: Default::default(),
            by_name: Default::default(),
            by_alias: Default::default(),
            next_id: 1.into(),
        }
    }
}

impl<T> CatalogEntrySet<T> {
    fn add(&mut self, name: &str, aliases: &[&str], info: T) -> Result<EntryId, CatalogError> {
        let mut errors = vec![];
        let name = UniCase::from(name);
        let aliases: Vec<UniCase<String>> = aliases
            .iter()
            .map(|a| UniCase::from((*a).to_string()))
            .collect();

        for a in &aliases {
            if self.by_alias.contains_key(a) {
                errors.push(CatalogErrorKind::EntryExists(a.as_ref().to_string()));
            }
        }

        if self.by_name.contains_key(&name) {
            errors.push(CatalogErrorKind::EntryExists(name.to_string()));
        }

        let id = self.next_id.fetch_add(1, Ordering::SeqCst).into();

        if let Some(_old_val) = self.entries.insert(id, info) {
            errors.push(CatalogErrorKind::Unknown);
        }

        match errors.is_empty() {
            true => {
                self.by_name.insert(name, id);

                for a in aliases {
                    self.by_alias.insert(a, id);
                }

                Ok(id)
            }
            _ => Err(CatalogError::new(errors)),
        }
    }

    fn find_by_id(&self, eid: &EntryId) -> Option<(EntryId, &T)> {
        self.entries.get(eid).map(|e| (*eid, e))
    }

    fn find_by_name(&self, name: &str) -> Option<(EntryId, &T)> {
        let name = UniCase::from(name);
        let eid = self.by_name.get(&name).or(self.by_alias.get(&name));

        eid.and_then(|eid| self.find_by_id(eid))
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn todo() {}
}
