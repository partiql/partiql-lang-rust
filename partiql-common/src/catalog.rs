#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct CatalogId(u64);

impl From<u64> for CatalogId {
    fn from(value: u64) -> Self {
        CatalogId(value)
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct EntryId(u64);

impl From<u64> for EntryId {
    fn from(value: u64) -> Self {
        EntryId(value)
    }
}

#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
pub struct ObjectId {
    catalog_id: CatalogId,
    entry_id: EntryId,
}

impl ObjectId {
    pub fn new(catalog_id: CatalogId, entry_id: EntryId) -> Self {
        Self {
            catalog_id,
            entry_id,
        }
    }

    pub fn catalog_id(&self) -> CatalogId {
        self.catalog_id
    }
    pub fn entry_id(&self) -> EntryId {
        self.entry_id
    }
}

impl From<(CatalogId, EntryId)> for ObjectId {
    fn from((catalog_id, entry_id): (CatalogId, EntryId)) -> Self {
        ObjectId::new(catalog_id, entry_id)
    }
}
