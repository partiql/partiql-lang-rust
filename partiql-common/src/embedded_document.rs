use dyn_clone::DynClone;
use dyn_hash::DynHash;
use std::fmt::Debug;
use std::sync::Arc;

#[derive(Clone)]
pub struct LazyEmbeddedDocument {
    bytes: Arc<Vec<u8>>,
    typ: DynEmbeddedTypeTag,
}

impl LazyEmbeddedDocument {
    pub fn new<B: Into<Vec<u8>>, T: Into<DynEmbeddedTypeTag>>(bytes: B, typ: T) -> Self {
        let bytes = Arc::new(bytes.into());
        let typ = typ.into();
        Self { bytes, typ }
    }
}

// dyn

pub type DynEmbeddedTypeTag = Box<dyn DynEmbeddedDocumentType>;

pub trait DynEmbeddedDocumentType: DynClone {
    fn construct(&self, bytes: &[u8]) -> Box<dyn EmbeddedDocument>;
}

dyn_clone::clone_trait_object!(DynEmbeddedDocumentType);

pub trait DynEmbeddedDocumentTypeFactory {
    fn to_dyn_type_tag(self) -> DynEmbeddedTypeTag;
}

// typed

pub type EmbeddedTypeTag<D> = Box<dyn EmbeddedDocumentType<Doc = D>>;
pub trait EmbeddedDocumentType: Clone {
    type Doc: EmbeddedDocument + 'static;

    fn construct(&self, bytes: &[u8]) -> Self::Doc;
}

#[cfg_attr(feature = "serde", typetag::serde)]
pub trait EmbeddedDocument: Debug + DynHash + DynClone {}

dyn_hash::hash_trait_object!(EmbeddedDocument);
dyn_clone::clone_trait_object!(EmbeddedDocument);

impl<T, D> DynEmbeddedDocumentType for T
where
    T: EmbeddedDocumentType<Doc = D>,
    D: EmbeddedDocument + 'static,
{
    fn construct(&self, bytes: &[u8]) -> Box<dyn EmbeddedDocument> {
        Box::new(EmbeddedDocumentType::construct(self, bytes))
    }
}

impl<T, D> DynEmbeddedDocumentTypeFactory for T
where
    T: EmbeddedDocumentType<Doc = D> + 'static,
    D: EmbeddedDocument + 'static,
{
    fn to_dyn_type_tag(self) -> DynEmbeddedTypeTag {
        Box::new(self)
    }
}
