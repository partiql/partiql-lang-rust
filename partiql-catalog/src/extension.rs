use crate::catalog::Catalog;
use std::error::Error;
use std::fmt::Debug;
use thiserror::Error;

pub trait Extension: Debug {
    fn name(&self) -> String;

    fn load(&self, catalog: &mut dyn Catalog) -> Result<(), ExtensionResultError>;
}

pub type ExtensionError = Box<dyn Error>;

/// Errors in extension.
///
/// ### Notes
/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
#[derive(Error, Debug)]
#[non_exhaustive]
pub enum ExtensionResultError {
    /// Load error. Error loading the extension.
    #[error("Scan error: `{}`", .0)]
    LoadError(ExtensionError),

    /// Read error. This will generally terminate the read operation.
    #[error("Scan error: `{}`", .0)]
    ReadError(ExtensionError),

    /// Data error. Generally this will result in a `MISSING` in place of this data item.
    #[error("Data error: `{}`", .0)]
    DataError(ExtensionError),
}
