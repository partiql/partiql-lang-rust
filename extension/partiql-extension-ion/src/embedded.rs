use core::fmt;
use delegate::delegate;
use ion_rs::{
    AnyEncoding, ConversionOperationError, Element, ElementReader, IonInput, IonResult, IonSlice,
    IonType, OwnedSequenceIterator, Reader, Sequence,
};
use ion_rs_old::{IonError, IonReader};
use partiql_common::pretty::{pretty_surrounded_doc, PrettyDoc};
use partiql_value::datum::{
    Datum, DatumCategoryOwned, DatumCategoryRef, DatumCattt, DatumLowerResult,
};
use partiql_value::embedded_document::{
    DynEmbeddedTypeTag, EmbeddedDocResult, EmbeddedDocValueIntoIterator, EmbeddedDocument,
    EmbeddedDocumentType,
};
use partiql_value::{EmbeddedDoc, Value, ValueIntoIterator};
use peekmore::{PeekMore, PeekMoreIterator};
#[cfg(feature = "serde")]
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::cell::{Ref, RefCell, RefMut};
use std::collections::VecDeque;
use std::fmt::{Debug, Display, Formatter};
use std::hash::{Hash, Hasher};
use std::iter::Peekable;
use std::ops::{ControlFlow, Deref, DerefMut};
use std::rc::Rc;
use std::slice;
use std::slice::Iter;
use std::sync::Arc;
use thiserror::Error;

#[derive(Default, Debug, Copy, Clone)]
pub struct EmbeddedIonType {}
impl EmbeddedDocumentType for EmbeddedIonType {
    type Doc = EmbeddedIon;

    fn construct(&self, bytes: Vec<u8>) -> EmbeddedDocResult<Self::Doc> {
        EmbeddedIon::parse(bytes.into(), EmbeddedDocStreamType::SingleTLV).map_err(Into::into)
    }

    fn name(&self) -> &'static str {
        "ion"
    }
}

/// Errors in boxed Ion.
///
/// ### Notes
/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
#[derive(Error, Debug, Clone, PartialEq)]
#[non_exhaustive]
pub enum BoxedIonError {
    /// Ion Writer error.
    #[error("Expected a sequence, but was `{elt}`")]
    NotASequence { elt: Element },
}

pub type Result<T> = std::result::Result<T, BoxedIonError>;

pub struct ElementIterator<R: ElementReader> {
    reader: R,
}

impl<R: ElementReader> Iterator for ElementIterator<R> {
    type Item = IonResult<Element>;

    fn next(&mut self) -> Option<Self::Item> {
        self.reader.read_next_element().transpose()
    }
}

struct IonContext {
    reader: PeekMoreIterator<ElementIterator<Reader<AnyEncoding, Vec<u8>>>>,
}

impl IonContext {
    pub fn new(data: Vec<u8>) -> IonResult<Self> {
        let reader = Reader::new(AnyEncoding, data)?;
        let reader = ElementIterator { reader }.peekmore();
        Ok(Self { reader })
    }
}

pub type IonContextPtr = Rc<RefCell<IonContext>>;

// TODO [EMBDOC] does this serialization work?
pub struct EmbeddedIon {
    ctx: IonContextPtr,
    doc_type: RefCell<EmbeddedDocType>,
}

#[cfg(feature = "serde")]
impl Serialize for EmbeddedIon {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        todo!()
    }
}

#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for EmbeddedIon {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        todo!()
    }
}

impl Clone for EmbeddedIon {
    // TODO [EMBDOC] make clone of doc less expensive
    fn clone(&self) -> Self {
        Self {
            ctx: self.ctx.clone(),
            doc_type: self.doc_type.clone(),
        }
    }
}

impl Hash for EmbeddedIon {
    fn hash<H: Hasher>(&self, state: &mut H) {
        todo!("EmbeddedIon.hash")
    }
}

#[cfg_attr(feature = "serde", typetag::serde)]
impl EmbeddedDocument for EmbeddedIon {
    fn into_dyn_iter(self: Box<Self>) -> EmbeddedDocResult<EmbeddedDocValueIntoIterator> {
        let iter = self.try_into_iter()?;

        Ok(
            Box::new(iter.map(|d| Box::new(d) as Box<dyn EmbeddedDocument>))
                as EmbeddedDocValueIntoIterator,
        )
    }

    fn category<'a>(&'a self) -> DatumCategoryRef<'a> {
        todo!() //self.doc_type.borrow().category()
    }

    fn into_category(self: Box<Self>) -> DatumCategoryOwned {
        todo!()
    }
}

impl Debug for EmbeddedIon {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("EmbeddedIon").field(&self.doc_type).finish()
    }
}

impl Display for EmbeddedIon {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        std::fmt::Display::fmt(&self.force(), f)
    }
}

impl EmbeddedIon {
    pub fn new(doc: EmbeddedDocType, ctx: IonContextPtr) -> Self {
        let doc_type = RefCell::new(doc);
        Self { ctx, doc_type }
    }

    pub fn parse(data: Vec<u8>, expected: EmbeddedDocStreamType) -> IonResult<Self> {
        let ctx = Rc::new(RefCell::new(IonContext::new(data)?));
        Ok(Self::new(EmbeddedDocType::Unexamined(expected), ctx))
    }

    pub fn parse_unknown(data: Vec<u8>) -> IonResult<Self> {
        Self::parse(data, EmbeddedDocStreamType::Unknown)
    }
    pub fn parse_tlv(data: Vec<u8>) -> IonResult<Self> {
        Self::parse(data, EmbeddedDocStreamType::SingleTLV)
    }

    pub fn parse_stream(data: Vec<u8>) -> IonResult<Self> {
        Self::parse(data, EmbeddedDocStreamType::Stream)
    }

    #[inline]
    fn force(&self) -> RefMut<'_, EmbeddedValue> {
        let doc = self.doc_type.borrow_mut();
        RefMut::map(doc, |mut doc| match doc {
            EmbeddedDocType::Unexamined(expected) => {
                *doc = EmbeddedDocType::Forced(self.init_reader(*expected));
                match doc {
                    EmbeddedDocType::Unexamined(_) => {
                        unreachable!()
                    }
                    EmbeddedDocType::Forced(doc) => doc,
                }
            }
            EmbeddedDocType::Forced(doc) => doc,
        })
    }

    fn init_reader(&self, expected: EmbeddedDocStreamType) -> EmbeddedValue {
        let reader = &mut self.ctx.borrow_mut().reader;
        let expected = match expected {
            EmbeddedDocStreamType::Unknown => {
                if reader.peek_nth(1).is_some() {
                    EmbeddedDocStreamType::Stream
                } else {
                    EmbeddedDocStreamType::SingleTLV
                }
            }
            other => other,
        };
        match expected {
            EmbeddedDocStreamType::Unknown => {
                unreachable!()
            }
            EmbeddedDocStreamType::Stream => EmbeddedValue::Stream(),
            EmbeddedDocStreamType::SingleTLV => {
                let elt = reader.next().expect("ion value"); // TODO [EMBDOC]
                let elt = elt.expect("ion element"); // TODO [EMBDOC]
                if reader.peek().is_some() {
                    // TODO error on stream instead of TLV?
                }

                match elt.try_into_sequence() {
                    Err(err) => EmbeddedValue::Value(err.original_value()),
                    Ok(seq) => EmbeddedValue::Sequence(seq.into_iter()),
                }
            }
        }
    }

    fn try_into_iter(mut self) -> Result<EmbeddedIonIterator> {
        self.force();
        let EmbeddedIon { ctx, doc_type } = self;

        let inner = match doc_type.into_inner() {
            EmbeddedDocType::Unexamined(_) => unreachable!("handled by `force`"),
            EmbeddedDocType::Forced(EmbeddedValue::Stream()) => EmbeddedIterType::Stream(),
            EmbeddedDocType::Forced(EmbeddedValue::Value(elt)) => match elt.try_into_sequence() {
                Err(err) => {
                    // TODO [EMBDOC]
                    // We could error? But generally PartiQL coerces to a singleton collection...
                    //Err(BoxedIonError::NotASequence { elt }),
                    EmbeddedIterType::Sequence(Sequence::new([err.original_value()]).into_iter())
                }
                Ok(seq) => EmbeddedIterType::Sequence(seq.into_iter()),
            },
            EmbeddedDocType::Forced(EmbeddedValue::Sequence(seq)) => {
                EmbeddedIterType::Sequence(seq.into_iter())
            }
        }
        .into();

        Ok(EmbeddedIonIterator { ctx, inner })
    }
}

#[derive(Debug, Copy, Clone)]
enum EmbeddedDocStreamType {
    Unknown,
    Stream,
    SingleTLV,
}

#[derive(Debug, Clone)]
enum EmbeddedDocType {
    Unexamined(EmbeddedDocStreamType),
    Forced(EmbeddedValue),
}

#[derive(Debug)]
enum EmbeddedValue {
    Stream(),
    Value(Element),
    Sequence(OwnedSequenceIterator),
}

impl Clone for EmbeddedValue {
    fn clone(&self) -> Self {
        // TODO [EMBDOC]
        match self {
            EmbeddedValue::Stream() => {
                todo!("stream not cloneable? ")
            }
            EmbeddedValue::Value(val) => EmbeddedValue::Value(val.clone()),
            EmbeddedValue::Sequence(seq) => {
                todo!("clone for Seq")
            }
        }
    }
}

impl Display for EmbeddedValue {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        // TODO [EMBDOC]
        match self {
            EmbeddedValue::Stream() => {
                todo!("stream not displayable? ")
            }
            EmbeddedValue::Value(val) => std::fmt::Display::fmt(val, f),
            EmbeddedValue::Sequence(seq) => {
                todo!("display for Seq")
            }
        }
    }
}

impl Datum<Value> for EmbeddedIon {
    fn is_null(&self) -> bool {
        match self.force().deref() {
            EmbeddedValue::Value(elt) => elt.is_null(),
            EmbeddedValue::Stream() => false,
            EmbeddedValue::Sequence(_) => false,
        }
    }

    fn is_sequence(&self) -> bool {
        match self.force().deref() {
            EmbeddedValue::Value(elt) => elt.as_sequence().is_some(),
            EmbeddedValue::Stream() => true,
            EmbeddedValue::Sequence(_) => true,
        }
    }

    fn is_ordered(&self) -> bool {
        match self.force().deref() {
            EmbeddedValue::Value(_) => false,
            EmbeddedValue::Stream() => false, // TODO [EMBDOC] is a top-level stream ordered?
            EmbeddedValue::Sequence(_) => true,
        }
    }

    fn lower(self) -> DatumLowerResult<Value> {
        todo!("lower for EmbeddedIon")
    }
}

impl<'a> DatumCattt<'a> for EmbeddedDocType {
    fn category(&'a self) -> DatumCategoryRef<'a> {
        match self {
            EmbeddedDocType::Unexamined(_) => {
                todo!("TODO [EMBDOC]")
            }
            EmbeddedDocType::Forced(doc) => doc.category(),
        }
    }

    fn into_category(self) -> DatumCategoryOwned {
        todo!()
    }
}

impl<'a> DatumCattt<'a> for EmbeddedValue {
    fn category(&'a self) -> DatumCategoryRef<'a> {
        todo!()
    }

    fn into_category(self) -> DatumCategoryOwned {
        todo!()
    }
}

#[derive(Debug)]
enum EmbeddedIterType {
    Stream(),
    Sequence(OwnedSequenceIterator),
}

struct EmbeddedIonIterator {
    ctx: IonContextPtr,
    inner: RefCell<EmbeddedIterType>,
}

impl Iterator for EmbeddedIonIterator {
    type Item = EmbeddedIon;

    fn next(&mut self) -> Option<Self::Item> {
        let elt = match self.inner.borrow_mut().deref_mut() {
            EmbeddedIterType::Stream() => {
                let elt = self.ctx.borrow_mut().deref_mut().reader.next();
                let elt = elt.transpose().expect("ion not error"); // TODO [EMBDOC]
                elt
            }
            EmbeddedIterType::Sequence(seq) => seq.next(),
        };
        elt.map(|elt| {
            EmbeddedIon::new(
                EmbeddedDocType::Forced(EmbeddedValue::Value(elt)),
                self.ctx.clone(),
            )
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ion_rs::{ElementReader, WriteAsIon};
    use std::io::Read;

    fn flatten_dump(doc: EmbeddedIon) {
        if doc.is_sequence() {
            for c in doc.try_into_iter().expect("TODO [EMBDOC]") {
                flatten_dump(c)
            }
        } else {
            println!("{:?}", doc);
        }
    }

    fn dump(data: Vec<u8>, expected_embedded_doc_type: EmbeddedDocStreamType) {
        println!("\n===========\n");

        let doc =
            EmbeddedIon::parse(data, expected_embedded_doc_type).expect("embedded ion create");

        flatten_dump(doc);
    }

    #[test]
    fn simple() {
        let one_elt: Vec<u8> =
            "[0, {a: 1, b:2, c: [], d: foo::(SYMBOL 3 2 1 {})}, [1,2,3,4]]".into();
        let stream: Vec<u8> = "0 {a: 1, b:2, c: [], d: foo::(SYMBOL 3 2 1 {})} [1,2,3,4]".into();

        dump(one_elt.clone(), EmbeddedDocStreamType::SingleTLV);
        dump(one_elt, EmbeddedDocStreamType::Unknown);
        dump(stream.clone(), EmbeddedDocStreamType::Stream);
        dump(stream, EmbeddedDocStreamType::Unknown);
    }
}
