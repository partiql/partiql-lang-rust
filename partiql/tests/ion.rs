use crate::common::{eval_query_with_catalog, TestError};
use assert_matches::assert_matches;
use partiql_catalog::catalog::PartiqlCatalog;
use partiql_catalog::extension::Extension;
use partiql_common::pretty::ToPretty;
use partiql_eval::eval::Evaluated;
use partiql_eval::plan::EvaluationMode;
use partiql_extension_ion::embedded::EmbeddedIonType;
use partiql_extension_value_functions::PartiqlValueFnExtension;
use partiql_value::embedded_document::DynEmbeddedDocumentTypeFactory;
use partiql_value::{EmbeddedDoc, Value};

mod common;

#[track_caller]
#[inline]
pub fn eval(statement: &str, mode: EvaluationMode) -> Result<Evaluated, TestError<'_>> {
    let mut catalog = PartiqlCatalog::default();
    let ext = PartiqlValueFnExtension::default();
    ext.load(&mut catalog)?;

    eval_query_with_catalog(statement, &catalog, mode)
}

#[test]
fn ion_simple() {
    let query = "select x from `(1 hi::2)` as x";
    // << {x: `1`}, {x: `hi::2`}>>

    let res = eval(query, EvaluationMode::Permissive);
    assert_matches!(res, Ok(_));
    let result = res.unwrap().result;

    insta::assert_snapshot!(result.to_pretty_string(25).expect("pretty"));
}

#[test]
fn ion_paths() {
    let query = "select x[1].foo from `([{foo:1}, {foo: 2}] ({foo: hi::1} {foo: world::2}))` as x";
    // << {x: `2`}, {x: `world::2`}>>

    let res = eval(query, EvaluationMode::Permissive);
    assert_matches!(res, Ok(_));
    let result = res.unwrap().result;

    insta::assert_snapshot!(result.to_pretty_string(25).expect("pretty"));
}

#[test]
fn ion_iter() {
    let contents = "[1,2,3,4]";
    let ion_typ = EmbeddedIonType::default().to_dyn_type_tag();
    let value = Value::EmbeddedDoc(Box::new(
        EmbeddedDoc::new(contents, ion_typ).expect("doc ctor"),
    ));

    let items: Vec<_> = value.into_iter().collect();
    dbg!(&items);
    assert_eq!(items.len(), 4);
}
