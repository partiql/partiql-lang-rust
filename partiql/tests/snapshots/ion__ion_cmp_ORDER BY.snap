---
source: partiql/tests/ion.rs
expression: "result.to_pretty_string(25).expect(\"pretty\")"
---
[
  { 'x': `1` },
  { 'x': 2 },
  { 'x': `3` },
  { 'x': 4 },
  { 'x': `5` },
  { 'x': 6 }
]
